src/main/java/io/confluent/connect/jdbc/dialect/DatabaseDialect.java

```diff
@@ -0,0 +1,461 @@
+
+  /**
+   * Use the supplied {@link SchemaBuilder} to add a field that corresponds to the column with the
+   * specified definition.
+   *
+   * @param column  the definition of the column; may not be null
+   * @param builder the schema builder; may not be null
+   * @return the name of the field, or null if no field was added
+   */
+  String addFieldToSchema(ColumnDefinition column, SchemaBuilder builder);
+
+  /**
+   * Apply the supplied DDL statements using the given connection. This gives the dialect the
+   * opportunity to execute the statements with a different autocommit setting.
+   *
+   * @param connection the connection to use
+   * @param statements the list of DDL statements to execute
+   * @throws SQLException if there is an error executing the statements
+   */
+  void applyDdlStatements(Connection connection, List<String> statements) throws SQLException;
+
+  /**
+   * Build the INSERT prepared statement expression for the given table and its columns.
+   *
+   * @param table         the identifier of the table; may not be null
+   * @param keyColumns    the identifiers of the columns in the primary/unique key; may not be null
+   *                      but may be empty
+   * @param nonKeyColumns the identifiers of the other columns in the table; may not be null but may
+   *                      be empty
+   * @return the INSERT statement; may not be null
+   */
+  String buildInsertStatement(
+      TableId table,
+      Collection<ColumnId> keyColumns,
+      Collection<ColumnId> nonKeyColumns
+  );
+
+  /**
+   * Build the UPDATE prepared statement expression for the given table and its columns. Variables
+   * for each key column should also appear in the WHERE clause of the statement.
+   *
+   * @param table         the identifier of the table; may not be null
+   * @param keyColumns    the identifiers of the columns in the primary/unique key; may not be null
+   *                      but may be empty
+   * @param nonKeyColumns the identifiers of the other columns in the table; may not be null but may
+   *                      be empty
+   * @return the UPDATE statement; may not be null
+   */
+  String buildUpdateStatement(
+      TableId table,
+      Collection<ColumnId> keyColumns,
+      Collection<ColumnId> nonKeyColumns
+  );
+
+  /**
+   * Build the UPSERT or MERGE prepared statement expression to either insert a new record into the
+   * given table or update an existing record in that table Variables for each key column should
+   * also appear in the WHERE clause of the statement.
+   *
+   * @param table         the identifier of the table; may not be null
+   * @param keyColumns    the identifiers of the columns in the primary/unique key; may not be null
+   *                      but may be empty
+   * @param nonKeyColumns the identifiers of the other columns in the table; may not be null but may
+   *                      be empty
+   * @return the upsert/merge statement; may not be null
+   */
```

Comment:
It's not caught, so using upsert with the general dialect will result in a failure, and this is the same behavior as before. Basically, you can't use upsert without a concrete dialect that overrides it.

I didn't add it to the JavaDoc because I thought it implied that it makes the implementation optional, when it really isn't optional if you want to use upsert.

However, I'll add some logic where this method is used to report a more meaningful exception.